src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,272 @@
+            }
+          }
+        }
+
+        JsonArray unservicedRequestsArray = responseJSON.get(ServiceNowConstants.UNSERVICED_REQUESTS).getAsJsonArray();
+
+        if (unservicedRequestsArray.size() > 0) {
+          // Add failed request Id to unserviced requests
+          if (failedRequestId != null) {
+            unservicedRequestsArray.add(failedRequestId);
+          }
+
+          // Process unserviced requests array into unserviced requests map
+          Map<String, RestRequest> unservicedRequestsMap = processUnservicedRequestsArray(restRequestsMap,
+                                                                                          unservicedRequestsArray);
+          // Retry unserviced requests
+          createPostRequest(unservicedRequestsMap, accessToken);
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while connecting to ServiceNow", e.getMessage());
+      throw new RetryableException();
+    } catch (Exception e) {
+      throw e;
+    }
+  }
+
+  private String getErrorMessage(String responseBody) {
+    try {
+      JsonObject jo = gson.fromJson(responseBody, JsonObject.class);
+      return jo.getAsJsonObject(ServiceNowConstants.ERROR).get(ServiceNowConstants.MESSAGE).getAsString();
+    } catch (Exception e) {
+      return e.getMessage();
+    }
+  }
+
+  private ServiceNowBatchRequest getPayloadRequest(Map<String, RestRequest> restRequests) {
+    ServiceNowBatchRequest payloadRequest = new ServiceNowBatchRequest();
+    payloadRequest.setBatchRequestId(UUID.randomUUID().toString());
+    payloadRequest.setRestRequests(new ArrayList<>(restRequests.values()));
+    return payloadRequest;
+  }
+
+  /**
+   * Process the unserviced requests array into an unserviced requests map.    .
+   *
+   * @param restRequestsMap         The map of rest requests
+   * @param unservicedRequestsArray An array of unserviced requests
+   * @return unservicedRequestsMap The map of unserviced requests
+   */
+  private Map<String, RestRequest> processUnservicedRequestsArray(Map<String, RestRequest> restRequestsMap,
+                                                                  JsonArray unservicedRequestsArray) {
+    Map<String, RestRequest> unservicedRequestsMap = new HashMap<>();
+    for (int i = 0; i < unservicedRequestsArray.size(); i++) {
+      unservicedRequestsMap.put(unservicedRequestsArray.get(i).getAsString(),
+                                restRequestsMap.get(unservicedRequestsArray.get(i).getAsString()));
+    }
+    return unservicedRequestsMap;
+  }
+
+  /**
+   * Retries to insert/update the records into ServiceNow table when RetryableException is thrown          .
+   *
+   * @param restRequestsMap The map of rest Requests
+   */
+  public void createPostRequestRetryableMode(Map<String, RestRequest> restRequestsMap) throws ExecutionException,
+    RetryException {
+    // Reset request counter
+    atomicCounter.set(1);
```

Comment:
why do we need to reset the counter? Any reason why we need incrementing ids instead of just using a UUID like you did in ServiceNowBatchRequest.setBatchRequestId()?

Reply:
The counter was being reset to re-initialize the starting request in a batch to 1.  We're using incrementing ids when we had `restRequestsLists`, as it was easier to process the unserviced requests. 

Since we are using `restRequestsMap` now, the incrementing ids have been replaced by UUID. 0a1c20a052aca8be095d4e1cde499b71fed53861